### PR TITLE
Uses DP/EMR instead of table scans

### PR DESCRIPTION
**Pull request by cantenesse**

### CODE REVIEW

File: base.yml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+aws:
```
Comment (kevinlr5) [inline base.yml]: is this for tests?

**Comment by cantenesse** [comment]
@kevinlr5 I'd like to submit this to the public Python package repo.  I don't think it's a big deal (given the public repo) and it will simplify the way we distribute.  In your review, can you be mindful of anything socotra specific?

File: base.yml (hunk 1 above)
Comment (cantenesse) [inline base.yml]: Yes.

File: bin/dynomite [hunk 2]
```diff
@@ -0,0 +1,163 @@
+#!/usr/bin/env python
+import argparse
+from dynomite import *
```
Comment (kevinlr5) [inline bin/dynomite]: is this how the boto stuff gets in?

File: dynomite/__init__.py [hunk 3]
```diff
@@ -1 +1,551 @@
...
+    return True
+
+
+def __copy_backup_from_source__(s3_conn, region, table,
+                                backup, src_bucket_name, debug):
+    logger = set_logging(level=debug)
+
+    tmp_bucket_name = str(uuid.uuid4())
+
+    dst_bucket = s3_conn.create_bucket(tmp_bucket_name, location='EU')
```
Comment (cantenesse) [inline dynomite/__init__.py]: EU is hardcoded for now.  Probably just need an enum (maybe via dict?) to map user provided region to the expected string for the AWS interface.

File: bin/dynomite (hunk 2 above)
Comment (cantenesse) [inline bin/dynomite]: No boto stuff in the CLI.  The boto stuff is in the dynomite module.

File: dynomite/__init__.py [hunk 4]
```diff
@@ -1 +1,551 @@
...
+        aws_secret_access_key=secret_access_key_id
+    )
+
+
+def get_tables_from_stack_definition(stack_definition):
+    tables = []
+    with open(stack_definition, 'r') as f:
+        config = yaml.load(f.read())
+
+    for table in config['aws']['dynamodb']:
```
Comment (kevinlr5) [inline dynomite/__init__.py]: this sort of thing can be written: if 'delete' in config['aws']['dynamodb'][table]

File: dynomite/__init__.py [hunk 5]
```diff
@@ -1 +1,551 @@
...
+    return tmp_bucket_name
+
+
+@retry(wait_exponential_multiplier=1000, wait_exponential_max=10000)
+def __check_pt_decreases__(conn, table):
+    t = dynamodb_conn.describe_table(table)
+    n_decreases_today = \
+        t['Table']['ProvisionedThroughput']['NumberOfDecreasesToday']
+
+    if n_decreases_today < 3:
```
Comment (kevinlr5) [inline dynomite/__init__.py]: this condition can be returned directly: return boolean expression

File: dynomite/__init__.py (hunk 4 above)
Comment (cantenesse) [inline dynomite/__init__.py]: Nice, will change that.

File: dynomite/__init__.py [hunk 6]
```diff
@@ -1 +1,551 @@
...
+    try:
+        with open('%s/templates/data_pipeline_'
+                  'default_role_policy.json' % (path), 'r') as f:
+            policy_document = f.read()
+        iam_conn.create_role('DataPipelineDefaultRole',
+                             assume_role_policy_document=policy_document)
+        logger.debug('Created role: DataPipelineDefaultRole')
+    except boto.exception.BotoServerError:
+        logger.error('Failed to create role: DataPipelineDefaultRole')
+        sys.exit()
```
Comment (kevinlr5) [inline dynomite/__init__.py]: I think this will exit with a 0 status

File: setup.py [hunk 7]
```diff
@@ -12,12 +11,22 @@
...
+                     'templates/restore_pipeline_def.json',
+                     'templates/data_pipeline_default_resource_role.json',
+                     'templates/data_pipeline_default_role.json',
+                     'templates/data_pipeline_default_resource_role_policy.json',
+                     'templates/data_pipeline_default_role_policy.json',
+                     'templates/user_policy.json']},
       url='https://www.socotra.com',
       maintainer='Chris Antenesse',
       maintainer_email='chris.antenesse@socotra.com',
       license='all rights reserved',
```
Comment (kevinlr5) [inline setup.py]: socotraness

Comment (cantenesse) [inline setup.py]: I think we're OK with this, no?

Comment (kevinlr5) [inline setup.py]: I'm cool, just looking for socotra related things

**Pull request merged**